2.10.2 Release Note (#262)
apache/pulsar-site: site2/website-next/release-notes/versioned/pulsar-2.10.2.md

---
id: pulsar-2.10.2
title: Apache Pulsar 2.10.2
sidebar_label: Apache Pulsar 2.10.2
---

#### 2022-10-24
### Important notice
- [fix][admin] Fix get non-persistent topics issue in Namespaces. [#16170](https://github.com/apache/pulsar/pull/16170)

### Broker
- [fix][broker] ManagedLedger: move to FENCED state in case of BadVersionException [#17736](https://github.com/apache/pulsar/pull/17736)
- [fix][broker] Issue 17588: Allow deletion of a namespace that was left in deleted status [#17592](https://github.com/apache/pulsar/pull/17592)

> - [fix][broker] Fix bookeeper packages npe [#17291](https://github.com/apache/pulsar/pull/17291)

Reviewer: ```suggestion
- [fix][broker] Fix BookKeeper packages npe [#17291](https://github.com/apache/pulsar/pull/17291)
```

- [fix][broker] Fix rack awareness cache expiration data race [#16825](https://github.com/apache/pulsar/pull/16825)
- [fix][broker] Fix stats-internal with option -m cause active ledger recover then close [#16662](https://github.com/apache/pulsar/pull/16662)
- [fix][broker] Fix deadlock in key-shared dispatcher [#16660](https://github.com/apache/pulsar/pull/16660)
- [fix][broker] PulsarLedgerManager: add missed return statement [#16607](https://github.com/apache/pulsar/pull/16607)

> - [fix][broker] Ensure ack-timeout task gets re-scheduled when there are exception in the final stage [#16337](https://github.com/apache/pulsar/pull/16337)

Reviewer 2: ```suggestion
- [fix][broker] Ensure the ack-timeout task gets re-scheduled when there is an exception in the final stage [#16337](https://github.com/apache/pulsar/pull/16337)
```

Reviewer 3: No comments. All good.

- [fix][broker] Fix schema does not replicate successfully [#17049](https://github.com/apache/pulsar/pull/17049)
- [fix][broker] Fix the replicator unnecessary get schema request for BYTES schema [#17523](https://github.com/apache/pulsar/pull/17523)
- [fix][broker] Fixed error when delayed messages trackers state grows to >1.5GB [#16490](https://github.com/apache/pulsar/pull/16490)
- [fix][broker] Increment topic stats outbound message counters and update rate after messages have been written to the TCP/IP connection [#17043](https://github.com/apache/pulsar/pull/17043)
- [fix][broker] ManagedCursor: mark delete no callback when create meta-ledger fail [#16841](https://github.com/apache/pulsar/pull/16841)
- [fix][broker] Multiple consumer dispatcher stuck when `unackedMessages` greater than `maxUnackedMessages` [#17483](https://github.com/apache/pulsar/pull/17483)
- [fix][broker] PulsarLedgerManager to pass correct error code to BK client [#16857](https://github.com/apache/pulsar/pull/16857)
- [fix][broker] Release the entry in getEarliestMessagePublishTimeOfPos. [#16386](https://github.com/apache/pulsar/pull/16386)

> - [fix][broker] Retry to delete the namespace if new topics created during the namespace deletion [#16676](https://github.com/apache/pulsar/pull/16676)

Reviewer: ```suggestion
- [fix][broker] Retry to delete the namespace if new topics are created during the namespace deletion [#16676](https://github.com/apache/pulsar/pull/16676)
```

- [fix][broker] Retry when DistributedIdGenerator has BadVersion error [#16491](https://github.com/apache/pulsar/pull/16491)
- [fix][broker] Skip reading more entries for a pending read with no more entries [#16400](https://github.com/apache/pulsar/pull/16400)
- [fix][broker] Support loadBalancerSheddingIntervalMinutes dynamic configuration [#16408](https://github.com/apache/pulsar/pull/16408)
- [fix][broker] The configuration loadBalancerNamespaceMaximumBundles is invalid [#16552](https://github.com/apache/pulsar/pull/16552)

> - [fix][broker] Topic policy reader can't recover when get any exception. [#17562](https://github.com/apache/pulsar/pull/17562)

Reviewer 2: ```suggestion
- [fix][broker] Topic policy reader can't recover when getting any exception. [#17562](https://github.com/apache/pulsar/pull/17562)
```

- [fix][broker] Unregister topic policy listener if managed ledger close failed [#17652](https://github.com/apache/pulsar/pull/17652)
- [fix][broker] Upgrade log4j2 version to 2.18.0 [#16884](https://github.com/apache/pulsar/pull/16884)
- [fix][broker] fix No such ledger exception [#16420](https://github.com/apache/pulsar/pull/16420)
- [fix][broker] fix broker unackmessages become a negative number [#17003](https://github.com/apache/pulsar/pull/17003)
- [fix][broker] remove exception log when access status.html [#17025](https://github.com/apache/pulsar/pull/17025)

> - [fix][broker] Consumer can't consume messages because there has two sames topics in one broker [#17526](https://github.com/apache/pulsar/pull/17526)

Reviewer 2: ```suggestion
- [fix][broker] Consumer can't consume messages because there are two same topics in one broker [#17526](https://github.com/apache/pulsar/pull/17526)
```

- [fix][broker] Fix getInternalStats occasional lack of LeaderInfo again [#16238](https://github.com/apache/pulsar/pull/16238)
- [fix][broker] Fix subscribe dispatcher limiter not be initialized [#16175](https://github.com/apache/pulsar/pull/16175)
- [fix][broker] Fix topic policies update not check message expiry [#15941](https://github.com/apache/pulsar/pull/15941)
- [fix][broker] Fix topic-level replicator rate limiter not init [#15825](https://github.com/apache/pulsar/pull/15825)
- [fix][broker] ManagedLedger metrics fail cause of zero period [#17257](https://github.com/apache/pulsar/pull/17257)
- [fix][broker] Prevent `StackOverFlowException` in SHARED subscription [#16968](https://github.com/apache/pulsar/pull/16968)
- [fix][broker] Fix message ack rate [#16108](https://github.com/apache/pulsar/pull/16108)
- [fix][broker] Fix bug: fail to expose managed ledger client stats to prometheus if bookkeeperClientExposeStatsToPrometheus is true [#16219](https://github.com/apache/pulsar/pull/16219)

> - [fix][broker] Fix npe when invoke replaceBookie. [#16239](https://github.com/apache/pulsar/pull/16239)

Reviewer: ```suggestion
- [fix][broker] Fix npe when invoking replaceBookie. [#16239](https://github.com/apache/pulsar/pull/16239)
```

- [fix][broker] Fail to update partition meta of topic due to ConflictException: subscription already exists for topic [#17488](https://github.com/apache/pulsar/pull/17488)
- [fix][broker] Fix NPE when ledger id not found in `OpReadEntry` [#15837](https://github.com/apache/pulsar/pull/15837)
- [Fix][broker] Fix topic dispatch rate limiter not init on broker-level [#16084](https://github.com/apache/pulsar/pull/16084)
- [opt][broker] Optimize topic policy with HierarchyTopicPolicies about replicatorDispatchRate [#14161](https://github.com/apache/pulsar/pull/14161)
- [fix][broker] Fix the broker close hanged issue. [#17689](https://github.com/apache/pulsar/pull/17689)
- [fix][broker] Fix topic dispatch rate limiter not init on broker-level #16084 [#17000](https://github.com/apache/pulsar/pull/17000)
- [fix][broker] Prevent StackOverFlowException in SHARED subscription [#17053](https://github.com/apache/pulsar/pull/17053)
- [fix][broker] skip mis-configured resource usage(>100%) in load balancer [#16937](https://github.com/apache/pulsar/pull/16937)

> - [fix][broker] Fix wrong unit of NIC speed on linux [#15304](https://github.com/apache/pulsar/pull/15304)

Reviewer: ```suggestion
- [fix][broker] Fix wrong unit of NIC speed on Linux [#15304](https://github.com/apache/pulsar/pull/15304)
```

- [fix][broker] Fix RawReader hasMessageAvailable returns true when no messages [#16443](https://github.com/apache/pulsar/pull/16443)
- [improve][broker] Expose topic level storage write and read rate metrics [#16855](https://github.com/apache/pulsar/pull/16855)
- [improve][broker] Broker/EntryFilter: make the delay for RESCHEDULED messages configurable (dispatcherFilterRescheduledMessageDelay) [#16602](https://github.com/apache/pulsar/pull/16602)
- [improve][broker] Use OrderedExecutor instead of OrderedScheduler for consumer dispatch [#16115](https://github.com/apache/pulsar/pull/16115)

> - [improve][broker] Avoid go through all the consumers to get the message ack owner [#16245](https://github.com/apache/pulsar/pull/16245)

Reviewer 2: ```suggestion
- [improve][broker] Avoid going through all the consumers to get the message ack owner [#16245](https://github.com/apache/pulsar/pull/16245)
```

- [fix][metadata] Set revalidateAfterReconnection true for certain failures [#17664](https://github.com/apache/pulsar/pull/17664)
- [improve][metadata] Update namespace policies would cause metadata-store thread waiting too long [#16438](https://github.com/apache/pulsar/pull/16438)
- [fix][storage] Fix offload read handle NPE. [#17478](https://github.com/apache/pulsar/pull/17478)
- [fix][storage] Fix OpAddEntry release error when exception in ManagedLedgerInterceptor [#17394](https://github.com/apache/pulsar/pull/17394)
- [fix][tableview] Fixed ack failure in ReaderImpl due to null messageId [#17828](https://github.com/apache/pulsar/pull/17828)

### Transaction
- [fix][txn] Cmd-Subscribe and Cmd-Producer will not succeed even after 100 retries [#16248](https://github.com/apache/pulsar/pull/16248)
- [fix][txn] Set TC state is Ready after open MLTransactionMetadataStore completely. [#13957](https://github.com/apache/pulsar/pull/13957)
- [fix][txn] Ack the same batch message different batchIndex with transaction [#16032](https://github.com/apache/pulsar/pull/16032)

> - [fix][txn] Allow producer enable send timeout in transaction [#16519](https://github.com/apache/pulsar/pull/16519)

Reviewer: ```suggestion
- [fix][txn] Allow producer to enable send timeout in transaction [#16519](https://github.com/apache/pulsar/pull/16519)
```

- [fix][txn] Fix NPE when ack message with transaction at cnx = null  [#16142](https://github.com/apache/pulsar/pull/16142)
- [fix][txn] Fix TopicTransactionBuffer ledger apend marker throw ManagedLedgerAlreadyClosedException [#16265](https://github.com/apache/pulsar/pull/16265)
- [fix][txn] Fix append txn message is lower than lowWaterMark decrease PendingWriteOps [#16266](https://github.com/apache/pulsar/pull/16266)
- [fix][txn] Fix ack with txn compute ackedCount error [#17016](https://github.com/apache/pulsar/pull/17016)
- [fix][txn] Fix pattern sub filter transaction system topic [#16533](https://github.com/apache/pulsar/pull/16533)
- [fix][txn] Fix race in TransactionMetaStoreHandler [#16147](https://github.com/apache/pulsar/pull/16147)

### Pulsar IO and Pulsar Functions

> - [fix][connector] Fixed kafka source config for `pulsar-io-kafka` when `consumerConfigProperties`="" [#16731](https://github.com/apache/pulsar/pull/16731)

Reviewer 2: ```suggestion
- [fix][connector] Fix Kafka source config for `pulsar-io-kafka` when `consumerConfigProperties`="" [#16731](https://github.com/apache/pulsar/pull/16731)
```
Can you keep tense consistent, check all occurencies, and update them?

- [fix][function] Fix python instance not process zip file correctly [#16697](https://github.com/apache/pulsar/pull/16697)
- [fix][function] Fixed error when user starts with the pulsar functions local runner [#16565](https://github.com/apache/pulsar/pull/16565)
- [fix][function] Ensure InternalConfigurationData data model is compatible across different versions [#17690](https://github.com/apache/pulsar/pull/17690)

### Tiered Storage
- [fix][tiered-storage] Don't cleanup data when offload met Metastore exception [#17512](https://github.com/apache/pulsar/pull/17512)
- [fix][tiered-storage] Fix the wrong secret key name get from env [#15814](https://github.com/apache/pulsar/pull/15814)
- [fix][tiered-storage] move the state check forward [#17020](https://github.com/apache/pulsar/pull/17020)
- [fix][common] Fix setManagedLedgerOffloadedReadPriority not work. [#16436](https://github.com/apache/pulsar/pull/16436)

### Pulsar SQL
- [fix][sql] Fix `messageQueue` release message issue. [#16155](https://github.com/apache/pulsar/pull/16155)

> - [fix][sql] Fix presto sql avro decode error when publish non-batched msgs [#17093](https://github.com/apache/pulsar/pull/17093)

Reviewer 2: ```suggestion
- [fix][sql] Fix Presto SQL Avro decode error when publish non-batched msgs [#17093](https://github.com/apache/pulsar/pull/17093)
```

> - [fix][sql] Fix Presto SQL Avro decode error when publish non-batched msgs [#17093](https://github.com/apache/pulsar/pull/17093)

Reviewer: ```suggestion
- [fix][sql] Fix Presto SQL Avro decode error when publishing non-batched msgs [#17093](https://github.com/apache/pulsar/pull/17093)
```

### Pulsar Proxy
- [fix][proxy] Do not preserve host when forwarding admin requests. [#16342](https://github.com/apache/pulsar/pull/16342)

> - [fix][proxy] Fix client service url [#16834](https://github.com/apache/pulsar/pull/16834)

Reviewer 2: ```suggestion
- [fix][proxy] Fix client service URL [#16834](https://github.com/apache/pulsar/pull/16834)
```

- [improve][authentication] Adapt basic authentication configuration with prefix [#16935](https://github.com/apache/pulsar/pull/16935)
- [improve][authentication] Improve get the basic authentication config [#16526](https://github.com/apache/pulsar/pull/16526)
- [improve][authentication] Support for get token from HTTP params [#16986](https://github.com/apache/pulsar/pull/16986)

### Dependency & Library updates
- [fix][dependency] Exclude the Netty Reactive Stream from asynchttpclient [#16312](https://github.com/apache/pulsar/pull/16312)
- [fix][dependency] Removing log4j-1.2-api from dependencies [#15991](https://github.com/apache/pulsar/pull/15991)
- [cleanup][dependency] Remove redundant pulsar-zookeeper-utils module [#16258](https://github.com/apache/pulsar/pull/16258)

### CI & Test
- [fix][ci] Remove post-commit trigger for old release branches (2.10 and previous) [#17570](https://github.com/apache/pulsar/pull/17570)
- [fix][ci] Fix some OWASP dependency problems. [#16260](https://github.com/apache/pulsar/pull/16260)
- [fix][test] Fix flaky-test RackAwareTest.testRackUpdate [#16071](https://github.com/apache/pulsar/pull/16071)

> - [fix][test] Catch exception when update data in mockZookeeper [#16473](https://github.com/apache/pulsar/pull/16473)

Reviewer: ```suggestion
- [fix][test] Catch exception when updating data in mockZookeeper [#16473](https://github.com/apache/pulsar/pull/16473)
```

- [fix][flaky-test] PersistentFailoverE2ETest.testSimpleConsumerEventsWithPartition [#16493](https://github.com/apache/pulsar/pull/16493)
- [fix][flaky-test] Try to fix flaky test related to PersistentTopicTest.setup [#16383](https://github.com/apache/pulsar/pull/16383)
- [fix][flaky-test] Fix testSplitBundleForMultiTimes [#16562](https://github.com/apache/pulsar/pull/16562)
- [fix][flaky-test] PulsarFunctionLocalRunTest.testE2EPulsarFunctionLocalRunMultipleInstances [#16872](https://github.com/apache/pulsar/pull/16872)
- [fix][flaky-test] AdminApi2Test.testDeleteNamespace [#17157](https://github.com/apache/pulsar/pull/17157)
- [fix][flaky-test] BrokerInterceptorTest.testProducerCreation [#17159](https://github.com/apache/pulsar/pull/17159)
- [fix][flaky-test] ConsumedLedgersTrimTest [#17116](https://github.com/apache/pulsar/pull/17116)
- [fix][flaky-test] ManagedCursorMetricsTest.testCursorReadWriteMetrics [#17045](https://github.com/apache/pulsar/pull/17045)

### Others
- [fix][build] Duplicate entry when merging services [#17659](https://github.com/apache/pulsar/pull/17659)

> - [cleanup][owasp] Supress false positive netty-tcnative [#17282](https://github.com/apache/pulsar/pull/17282)

Reviewer 2: ```suggestion
- [cleanup][owasp] Suppress false positive netty-tcnative [#17282](https://github.com/apache/pulsar/pull/17282)
```
